### PR TITLE
[Quant][Inductor] Add qlinear_pointwise.binary op for X86Inductor backend

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122667
* #122593
* #122387
* __->__ #122288

**Description**
Add qlinear_binary op for X86Inductor backend of quantization PT2E. It only supports `add` and `add_relu` now.
It will use post op sum if the extra input has the same dtype as output. Otherwise, it uses binary add.
```
+-------------------+--------------+---------------+
| Extra input dtype | Output dtype | Post op       |
+-------------------+--------------+---------------+
| Fp32/bf16         | fp32/bf16    | sum or add*   |
+-------------------+--------------+---------------+
| Fp32/bf16         | int8         | add           |
+-------------------+--------------+---------------+
| int8              | fp32/bf16    | not supported |
+-------------------+--------------+---------------+
| int8              | int8         | sum           |
+-------------------+--------------+---------------+
*Use sum if extra input and output have the same dtype; otherwise use add.
```

**Test plan**
python test_quantization.py -k test_qlinear_add_pt2e
python test_quantization.py -k test_qlinear_add_relu_pt2e

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by leslie-fang-intel — APPROVED** [review]
LGTM

**Comment by leslie-fang-intel** [comment]
Refer to the comment in https://github.com/pytorch/pytorch/pull/122387#discussion_r1538517985, we should also need linear-add with post op binary add since the linear add recipe doesn't have quant node between add and its extra input node.

File: aten/src/ATen/native/quantized/library.cpp [hunk 1]
```diff
@@ -267,4 +267,7 @@ TORCH_LIBRARY(onednn, m) {
   // Linear with unary postop
   m.def(TORCH_SELECTIVE_SCHEMA("onednn::qlinear_pointwise(Tensor qx, float x_scale, int x_zero_point, Tensor qw, Tensor w_scale, Tensor w_zero_point, Tensor? bias, float output_scale, int output_zero_point, ScalarType? output_dtype, str post_op_name, Scalar?[] post_op_args, str post_op_algorithm) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("onednn::qlinear_pointwise.tensor(Tensor qx, Tensor x_scale, Tensor x_zero_point, Tensor qw, Tensor w_scale, Tensor w_zero_point, Tensor? bias, float output_scale, int output_zero_point, ScalarType? output_dtype, str post_op_name, Scalar?[] post_op_args, str post_op_algorithm) -> Tensor"));
+  // Linear with unary postop
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/library.cpp:270]: Thanks. It's fixed.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 2]
```diff
@@ -979,22 +1008,56 @@ static at::Tensor linear_int8_with_onednn_weight(
...
-  dnnl::algorithm post_op_algo = dnnl::algorithm::undef;
-  if (post_op_name == "gelu") {
-    if (post_op_algorithm == "none") {
-      post_op_algo = dnnl::algorithm::eltwise_gelu_erf;
-    } else if (post_op_algorithm == "tanh") {
-      post_op_algo = dnnl::algorithm::eltwise_gelu_tanh;
+  // Get op attr for primitive
+  // Note: output_scale & output_zero_point are for re-quantization of the final output.
+  // And accum_scale & accum_zero_point are for dequantization of accum.
+  auto op_attr = [&]() -> ideep::attr_t {
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:1019]: The util function was intended to serve both qlinear and qconv. However, qconv didn't use this function and the logic and supported fusions are different from that of qlinear. Besides, the util function requires more arguments to support binary post op, which does not look concise enough. So, I removed the util function and put it here.

[the remark this answers is not in the document]
Comment (jgong5) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:1019]: Just curious why and how qconv post-op fusion is different from qlinear's. I don't see any linear specific logic in your code...

[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:1019]: Ok. I will try to merge them.

**Comment by Xia-Weiwen** [comment]
> Refer to the comment in [#122387 (comment)](https://github.com/pytorch/pytorch/pull/122387#discussion_r1538517985), we should also need linear-add with post op binary add since the linear add recipe doesn't have quant node between add and its extra input node.

Ok. I have added the support for binary add

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: I have added the util function back and used it in both qlinear and qconv implementation.

**Comment by Xia-Weiwen** [comment]
Removed the constraint that binary add can only be used for int8 output so that we can support the following pattern with binary add.
```
bf16 linear (x)      fp32 extra input (y)
             \           /
               x.add_(y)
```

**Pull request merged**